### DISCUSSION TITLE
[Proposal]: Equivalent to "throws" keyword

### DISCUSSION THREAD

**Discussion by tte18** [dotnet/csharplang#4059] [Language Ideas] [answered]
# Equivalent to throws keyword

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

In C# you do not really know whether can method throw exception or not. All you can do now is use method naming conventions (btw, I did not found any) like in dynamic languages such as Ruby. Which is a shame for such beautiful, static language as C#. What I suggest is to adopt Java's `throw` keyword equivalent.

## Motivation
[motivation]: #motivation

This will help to reduce errors in runtime.

## Detailed design
[design]: #detailed-design

In Java if method can throw exception you _must_ either catch it or specify, that this method can throw exception of some type, even if you do not use this method at all.
```java
class Main {
  public static void test() throws Exception {
    if (false) {
      throw new Exception();
    }
  }

  public static void main(String[] args) { }
}
```
It can be optional or mandatory.

Suggested C# syntax:
```csharp
class Program
{
  static void test() ! Exception
  {
    throw new Exception();
  }
        
  static void Main(string[] args)
  {
    Console.WriteLine("Hello World!");
  }
}
```

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by YairHalberstadt** [comment]
C# very intentionally did not adopt java's checked exceptions, which are also widely hated by Java developers - both scala and kotlin don't have them despite being made for the JVM for example.

I think you'll have to explain why that decision should be revisited, or come up with a design that avoids many of the pitfalls of Java's design for this to be considered.

A much more promising direction is to take a leaf out of functional languages, and introduce a `Result<T, TError>` type along with coroutines for working with it. The advantage of this approach is that it composes much better since it's just an ordinary type system construct.

**Comment by YairHalberstadt** [comment, marked as answer]
See here for why C# doesn't have checked exceptions https://www.artima.com/intv/handcuffs.html

**Comment by dstarkowski** [comment]
There's an `<exception/>` tag in xml docs. Maybe better solution would be to surface it in IDE? Of course the usual "you could write an analyzer" answer also applies.
